frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx

```diff
@@ -207,79 +50,155 @@ const DatasetRow: FC<Props> = ({
+  const name = dataset.name || file?.name || dataset.id;
+
+  // (thuang): We need to poll the collection until the name is populated,
+  // which indicates other metadata are populated too
+  useCheckCollectionPopulated({
     invalidateCollectionQuery,
-    isLoading,
-    lastUploadProgress,
-    nameIsPopulated,
+    isNamePopulated,
+    upload_progress,
+  });
+
+  const hasFailed = checkIfFailed(datasetStatus);
+  const { isFailed, error } = hasFailed;
+
+  useCancelDatasetStatusQuery({
+    datasetId: dataset.id,
+    isFailed,
+    isNamePopulated,
     queryCache,
-  ]);
+  });
+
+  useUploadProgress({
+    newUploadProgress: upload_progress,
+    setUploadProgress,
+    uploadProgress,
+  });
+
+  const isLoading = checkIfLoading(datasetStatus);
+
+  const {
+    tissue,
+    assay,
+    disease,
+    organism,
+    cell_count,
+  } = aggregateDatasetsMetadata([dataset]);
 
   return (
     <StyledRow>
       <DetailsCell>
         <TitleContainer>
           <Checkbox onChange={() => checkHandler(dataset.id)} />
           <div>{name}</div>
-          {errorTooltip(hasFailed)}
+          <ErrorTooltip isFailed={isFailed} error={error} />
         </TitleContainer>
-        {isLoading && renderUploadStatus(datasetStatus)}
+        {isLoading && (
+          <UploadStatus
+            isValidating={
+              datasetStatus.validation_status === VALIDATION_STATUS.VALIDATING
+            }
+            progress={datasetStatus.upload_progress}
+          />
+        )}
       </DetailsCell>
-      {conditionalPopover(tissue, isLoading, hasFailed.isFailed)}
-      {conditionalPopover(assay, isLoading, hasFailed.isFailed)}
-      {conditionalPopover(disease, isLoading, hasFailed.isFailed)}
-      {conditionalPopover(organism, isLoading, hasFailed.isFailed)}
-      {cellCount(cell_count, hasFailed.isFailed, isLoading)}
+      <Popover values={tissue} isLoading={isLoading} isFailed={isFailed} />
+      <Popover values={assay} isLoading={isLoading} isFailed={isFailed} />
+      <Popover values={disease} isLoading={isLoading} isFailed={isFailed} />
+      <Popover values={organism} isLoading={isLoading} isFailed={isFailed} />
+      <CellCount cellCount={cell_count} isLoading={isLoading} />
       <RightAlignedDetailsCell>
-        {nameIsPopulated && (
+        {isNamePopulated && (
           <Button intent={Intent.PRIMARY} outlined text="Explore" />
         )}
       </RightAlignedDetailsCell>
     </StyledRow>
   );
 };
 
+function useCheckCollectionPopulated({
+  invalidateCollectionQuery,
+  isNamePopulated,
+  upload_progress,
+}: {
+  invalidateCollectionQuery: () => void;
+  isNamePopulated: boolean;
+  upload_progress: number;
+}) {
+  useEffect(() => {
+    let intervalId: number | undefined = undefined;
+
+    if (!intervalId && upload_progress === 1 && !isNamePopulated) {
+      intervalId = window?.setInterval(() => {
+        if (upload_progress === 1 && !isNamePopulated) {
+          invalidateCollectionQuery();
```

Comment:
Instead of using `invalidateCollectionQuery` we might be better off calling the `refetch` function that `useQuery` returns

https://react-query-v2.tanstack.com/docs/api#usequery
![image](https://user-images.githubusercontent.com/8716829/104357805-82378f80-54c2-11eb-89dc-8c1b1124a8e0.png)